internal/tools/generator-schema-snapshot/main.go

```diff
@@ -0,0 +1,102 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+
+	. "github.com/dave/jennifer/jen"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/provider"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+)
+
+const (
+	SchemaPath = "github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+)
+
+func main() {
+	if len(os.Args) != 2 {
+		log.Fatal("Usage: generator-schema-snapshot <reource_type>")
+	}
+	rt := os.Args[1]
+	res, ok := provider.AzureProvider().ResourcesMap[rt]
+	if !ok {
+		log.Fatalf("unknown resource type %q", rt)
+	}
+
+	f := NewFile("main")
+	f.ImportName("github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk", "")
+
+	f.Var().Id("_").Op("=").Add(SchemaMap(res.Schema))
+
+	fmt.Printf("%#v", f)
+}
+
+func ResourceValue(res *pluginsdk.Resource) Dict {
+	return Dict{
+		Id("Schema"): SchemaMap(res.Schema),
+	}
+}
+
+func SchemaMap(m map[string]*pluginsdk.Schema) *Statement {
+	dict := Dict{}
+	for k, v := range m {
+		dict[Lit(k)] = Values(SchemaValue(v))
+	}
+	return Map(String()).Op("*").Qual(SchemaPath, "Schema").Values(dict)
+}
+
+func SchemaValue(sch *pluginsdk.Schema) Dict {
+	out := Dict{}
+
+	var t Code
+	switch sch.Type {
+	case pluginsdk.TypeBool:
+		t = Qual(SchemaPath, "TypeBool")
+	case pluginsdk.TypeInt:
+		t = Qual(SchemaPath, "TypeInt")
+	case pluginsdk.TypeFloat:
+		t = Qual(SchemaPath, "TypeFloat")
+	case pluginsdk.TypeString:
+		t = Qual(SchemaPath, "TypeString")
+	case pluginsdk.TypeList:
+		t = Qual(SchemaPath, "TypeList")
+	case pluginsdk.TypeMap:
+		t = Qual(SchemaPath, "TypeMap")
+	case pluginsdk.TypeSet:
+		t = Qual(SchemaPath, "TypeSet")
```

Comment:
This'll probably also need to account for/placeholder the custom hash functions too?

Reply:
The schema `Type` used here is only used when the state is stored in a legacy flatmap format:

```go
type StateUpgrader struct {
    ...
    // Type describes the schema that this function can upgrade. Type is
    // required to decode the schema if the state was stored in a legacy
    // flatmap format.
    Type cty.Type
```

Also, the `Set` hash function seems in only used for the provider when adding/removing elements to the set, serving the diff purpose. Whilst it is not used anywhere in the state migrate function for flatmap format, where the actual representation of the schema type is in `cty.Type`, whose Set type has no hash function at all?

Reply:
ultimately we're still going to need that context (at least for a while) after migrating to Framework though, since we're not planning on using Framework directly fwiw

Reply:
@tombuildsstuff All right, I've now add a `TODO` identifier if there is non nil `Set` function in the previous schema. This makes the generated schema fail to compile, to imply that users need to use the exact set function instead.